sync15-adapter/src/token.rs

```diff
@@ -0,0 +1,133 @@
+    }
+}
+
+fn token_url(base_url: &str) -> Result<Url> {
+    let mut url = Url::parse(base_url)?;
+    // kind of gross but avoids problems if base_url has a trailing slash.
+    url.path_segments_mut()
+        // We can't do anything anyway if this is the case.
+       .map_err(|_| error::unexpected("Bad tokenserver url (cannot be base)"))?
+       .extend(&["1.0", "sync", "1.5"]);
+    Ok(url)
+}
+
+impl TokenserverClient {
+    #[inline]
+    pub fn server_timestamp(&self) -> ServerTimestamp {
+        self.server_timestamp
+    }
+
+    #[inline]
+    pub fn token(&self) -> &TokenserverToken {
+        &self.token
+    }
+
+    pub fn new(request_client: &Client, base_url: &str, access_token: String, key_id: String) -> Result<TokenserverClient> {
+        let mut resp = request_client.get(token_url(base_url)?)
+                                     .header(Authorization(Bearer { token: access_token }))
+                                     .header(XKeyID(key_id))
+                                     .send()?;
+
+        if !resp.status().is_success() {
+            warn!("Non-success status when fetching token: {}", resp.status());
+            // TODO: the body should be JSON and contain a status parameter we might need?
+            debug!("  Response body {}", resp.text().unwrap_or("???".into()));
+            if let Some(seconds) = resp.headers().get::<RetryAfter>().map(|h| **h) {
+                bail!(error::ErrorKind::BackoffError(seconds));
+            }
+            bail!(error::ErrorKind::TokenserverHttpError(resp.status()));
+        }
+
+        let token: TokenserverToken = resp.json()?;
+
+        let timestamp = resp.headers()
+                            .get::<XTimestamp>()
+                            .map(|h| **h)
+                            .ok_or_else(|| error::unexpected(
+                                "Missing or corrupted X-Timestamp header from token server"))?;
+        let credentials = hawk::Credentials {
+            id: token.id.clone(),
+            key: hawk::Key::new(token.key.as_bytes(), hawk::Digest::sha256())?,
+        };
+        Ok(TokenserverClient {
+            token,
+            credentials,
+            server_timestamp: timestamp
+        })
+    }
+
+    pub fn authorization(&self, req: &Request) -> Result<Authorization<String>> {
+        let url = req.url();
+
+        let path_and_query = match url.query() {
+            None => Cow::from(url.path()),
+            Some(qs) => Cow::from(format!("{}?{}", url.path(), qs))
+        };
+
+        let host = url.host_str().ok_or_else(||
+            error::unexpected("Tried to authorize bad URL using hawk (no host)"))?;
+
+        // Known defaults exist for https? (among others), so this should be impossible
+        let port = url.port_or_known_default().ok_or_else(||
+            error::unexpected(
+                "Tried to authorize bad URL using hawk (no port -- unknown protocol?)"))?;
+
+        let header = hawk::RequestBuilder::new(
```

Comment:
Eventually, this could cache builders by URL, I imagine there are some routes we'll hit more frequently than others...